### ISSUE TITLE
Possible issue with parsing very long OpenShift annotations with serialized encapsulation

### ISSUE THREAD

**Issue by mmclanerh** [app-sre/qontract-reconcile#960]
**Problem:** Unable to update a SelectorSyncSet, with a large (over 20k characters) last-applied-configuration annotation.

Error message from Openshift API:
```
The nnn-objectType "nnn-objectName" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update
```

**Workaround:** Manually edited the (SelectorSyncSet) object and removed metadata.annotations.kubectl.kubernetes.io/last-applied-configuration on the cluster. This allowed the "oc apply" to apply successfully.

**_Possible_ Upstream Bug-ID(s):**
* https://github.com/kubernetes/kubernetes/issues/70674 (metadata.resourceVersion mismatch blocking PUT calls)

**Symptom(s):** Encountered an error from the OpenShift API where attempting to "oc apply" an object with a very large (20k character) metadata.annotations.kubectl.kubernetes.io/last-applied-configuration results in error. The annotation blob (which is serialized) converted to yaml, results in some new-lines following white space in the output, which may or may not be triggering this condition.

**Data Gathering, Other artifacts:**

When exporting the object as json (oc get nnType/nnObject -ojson), the line feeds are not present the the serialization is sane. When exporting the object as yaml (oc get nnType/nnObject -oyaml, the line feeds are present in some locations where whitespace is present, which seems to be corrupting the serialization.

Likewise, if I dump the entire object to json, and then attempt to convert the json to yaml using yq, the result is also new-line-included serialization -- so this may be exclusive to how our oc-apply the parse/re-parsing the serialized output and attempting to canonicalize [1]. Trying to track down the default line-wrap width there..

[1] https://github.com/app-sre/qontract-reconcile/blob/761f5aada45761c3f6b0d4058d2f555140ee1a09/utils/openshift_resource.py#L243-L260

Removing the single serialized line from json input demonstrates 8 extra new-lines are being added when parsed into yaml. 
```
$ yq -y .metadata.annotations[]  raw.json  | wc -l
14
$ yq -y .metadata.annotations[]  raw.last-removed.json  | wc -l
5

# json: 1 line, 20,247 characters
# yaml: 9 lines, 17,185 characters
$ jq . line.json  | wc
      1      25   20182
$ yq -y . line.json | wc 
      9      25   17185
```

It is also possible that some encoding within the yaml (de-)serialization is causing the syntax to become multi-lined or invalid. This size of annotation is also highly irregular and a-typical for openshift objects.

**Comment by mmclanerh** [comment]
@maorfr @jmelis  this is a wild one.. for internal reference, the object annotation exhibiting this is: SelectorSyncSet/cloud-ingress-operator  ||  hive-staging.

**Comment by mmclanerh** [comment]
I suspect the length maybe is a misnomer, and the real issue lies with a SelectorSyncSet and it's last-applied-configuration values when using oc apply, perhaps similar to:

https://github.com/kubernetes/kubernetes/issues/70674#issuecomment-519972588

**Comment by maorfr** [comment]
Jira: https://issues.redhat.com/browse/APPSRE-2064

**Comment by mmclanerh** [comment]
A suggested code workaround is to regurgitate an existing metadata.resourceVersion from an object into the to-be-applied object. Curious how this ends up triggering in the first place... and it's unclear to me if this could have existing/future OSD object/lifecycle behavior impacts vs obliterating the last-applied-configuration entirely.
```
if newResource.metadata.resourceVerision == "" {
  version := getOldResource.metadata.resourceVerision
  newResource.metadata.resourceVerision = version
}
```

**Comment by maorfr** [comment]
fixed by #982
